File: Debugging/Arithmetic/underflow.c
#include <fenv.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

int main() {
    float x[] = {
        1.0e-31f,
        1.0e-31f,
        5.0f,
        1.0e31f,
        1.0e31f
    };
    int i;
    float prod = 1.0f;
    float log_prod = 0.0f;
    double dprod = 1.0;
#ifdef FPE_TRAP
    feenableexcept(FE_DIVBYZERO | FE_INVALID | FE_OVERFLOW | FE_UNDERFLOW);
#endif
    for (i = 0; i < 5; i++)
        prod *= x[i];
    for (i = 0; i < 5; i++)
        dprod *= x[i];
    for (i = 0; i < 5; i++)
        log_prod += log(x[i]);
    printf("float product = %e\n", prod);
    printf("double product = %le\n", dprod);
    printf("log float product = %e\n", exp(log_prod));
    prod = 1.0f;
    for (i = 4; i >= 0; i--)
        prod *= x[i];
    printf("float product (reverse) = %e\n", prod);
    return EXIT_SUCCESS;
}

File: Debugging/Arithmetic/underflow.f90
program underflow

    use, intrinsic :: iso_fortran_env, only : sp => REAL32, dp => REAL64
    implicit none
    integer, parameter :: n = 5
    real(kind=sp), dimension(n) :: x = [ 1.0e-31_sp, 1.0e-31_sp, 5.0_sp, &
                                         1.0e31_sp, 1.0e31_sp ]
    integer :: i;
    real(kind=sp) :: prod
    real(kind=sp) :: log_prod
    real(kind=dp) :: dprod
    prod = 1.0_sp
    log_prod = 0.0_sp
    dprod = 1.0_dp
    do i = 1, n
        prod = prod*x(i)
    end do
    do i = 1, n
        dprod = dprod*x(i)
    end do
    do i = 1, n
        log_prod = log_prod + log(x(i))
    end do
    print '(A, F20.15)', 'REAL32 product = ', prod
    print '(A, F20.15)', 'REAL64 product = ', dprod
    print '(A, F20.15)', 'log REAL32 product = ', exp(log_prod)
    prod = 1.0_sp;
    do i = 1, n
        prod = prod*x(i)
    end do
    print '(A, F20.15)', 'REAL32 product (reverse) = ', prod

end program underflow
